### ISSUE TITLE
How to assign the atomic charges on-the-fly in the same time of molecular dynamics

### ISSUE THREAD

**Issue by yuxzhou** [lammps/lammps#3619]
**Summary**

I wish to change the values of atomic charges during a MD simulation, in which the atomic charges will be given by an external generator. I wonder is it possible to realise this in the current lammps package?

**LAMMPS Version and Platform**
I am using the latest version of LAMMPS, i.e., LAMMPS (23 Jun 2022 - Update 2), and running the simulations on Archer2.

**Details**

I am doing a MD simulation under an external electric field using the `fix efield` command. Instead of using a constant value for each atomic charge at different timesteps, I wish to assign variable atomic charges to the atoms at different timesteps. In this process, the variable atomic charges will be predicted by an external machine-learning model. I wonder is there any possible way that LAMMPS can reads in the external charges (predicted by the ML model) at each MD step and then do the MD simulation using the updated charges?

**Comment by akohlmey** [comment]
This is going to be a non-trivial process. But more importantly, there is some information missing about how you want to realize what you describe and how this ultimately needs to be implemented. Some questions:

- What are your potential functions or force field? For example, if you use a Kspace solver, then you will need to re-initialize the Kspace solver in every step since it typically pre-computes a bunch of factors under the assumption that charges do not change for efficiency reasons.
- How do you trigger the ML calculation? How would you access the data it produces? How would that work when running LAMMPS in parallel with MPI (across multiple compute nodes) ? How time consuming would the ML calculation be versus the force evaluation in LAMMPS?
- What is your experience in programming? What programming languages do you have the most experience with?

**Comment by yuxzhou** [comment]
Thanks very much for you reply!

Regarding your questions:

- I am using a machine-learned potential (based on Gaussian potential framework), viz. GAP potential, via the `ML-QUIP` package in LAMMPS. It only provides the energy and forces of the system at each step during the MD simulation. However, we are going to consider a more realistic scenario in electronic devices in which electric field is applied. So we plan to switch on the electric field in LAMMPS and assign atomic charges to each atom. We can definitely use a constant charge for each atom, but to make it more realistic, a variable charge will be appreciated.
- We are running MD simulations in NVT ensenble with the prediced energies and forces from a GAP potential (using the `QUIP` interface). It will be very tricky of how to do this on-the-fly charge in parallel with MPI. But we plan to fit another ML model, e.g., based on Gaussian process or neuron network, which can predict the charges according to its local atomic environments. I don't assume that LAMMPS has an interface to read a ML model for charge information?
- I have most experiences in Python, and some in Fortran. I'm thinking whether `ase` can be used to triger the MD simulations with a lammps interface and read the charge information? (I don't assume that `ase` can do MD simulations with applied electric field?)

**Comment by akohlmey** [comment]
Unfortunately, your answer prompt more questions than what they answer. This is also difficult for me to give specific advice, since my knowledge of machine learning is limited. Perhaps @athomps can chime in and correct and expand on what I a write below.

First off, I am not aware of any ML model that does *set* point charges. Rather, the Coulomb interaction contribution is already implicitly included. I am also aware of a concept of combining a machine learning model with explicit Coulomb interactions, which requires computing and subtracting the Coulomb contributions from the training data and then adding the classical Coulomb contributions through an overlay to that machine learning data. The purpose of this would be to combine the ML interaction model, which is inherently short ranged, with proper treatment of long-range electrostatics.

For your proposed model, I would doubt that using fix efield is suitable, since it would need to be included into the training data and the ML formalism as an additional degree of freedom. The reason for that is that an electrical field will induce a polarization in your material. If I understand your description correctly, you are trying to realize something like that, but are stopping half-way and rather than having the response to the external field enter your ML training and then apply it to the computed force and energy directly, want to separate that out and rather compute it after the fact.
To me this seems needlessly complicated. What you are effectively doing by exporting a per-atom charge is just adding a (simple, linear) force and energy term, so you could just add it within your ML computation. If I understand correctly, you would have to modify the ML code accordingly anyway. If you can make it predict a charge, you can just take that (plus the information about the field strength) and add a force term directly within the ML framework (or just include the field strength in the training and then infer the forces for "everything" in one go).

At the moment, I would be more concerned about the conceptual approach rather than the technical details. To me trying to send back charge data to LAMMPS only would make sense, if you want to do the *entire* Coulomb treatment from within LAMMPS (as I was referring to before). But that has other problems and complications for which LAMMPS has code e.g. in the ELECTRODE package. Using fix efield would imply using a slab configuration and then the usual long-range solvers used by LAMMPS (pppm, ewald) will not necessarily converge to meaningful results unless the field is very weak (which in turn would then result in effects that are smaller than the typical statistical fluctuations of the data for the achievable system sizes and length of trajectories)

**Comment by athomps** [comment]
As @akohlmey indicates, the actual code implementation is the least of your worries.  If your goal is to make something that is more realistic than fixed charges, whatever you construct will have to be trained on charge and energy data from some ground truth data source like DFT, and this will have to be done at the same time that the the GAP model is trained on energy.  If you attempt to train the charge model with a fixed GAP model, you will not be able to get consistent charges and energies.  The details of how an ML charge model should be constructed is an area of active research that would take us even farther away from the topic of the LAMMPS code, so I'll just say that ML methods for charge are much less mature than ML methods for energy.  If you are still serious about pursuing this, I suggest you first learn how to construct an ML energy model e.g. using a software package like FitSNAP [shameless plug].  Finally, if you want to ignore all of this advice and forge ahead with an ML charge model that is retrofitted to a GAP potential, I suggest you write a new LAMMPS `fix` style, similar to `fix qeq`.

**Comment by srtee** [comment]
From my experience with the ELECTRODE package, I agree with @akohlmey and @athomps that coding into LAMMPS will be the easiest part of the process. Scientific validity aside, on the LAMMPS end, you would proceed by writing a fix that reads in charges from your external source (presumably as a vector of doubles) and modifies the LAMMPS internal charge data at pre-force. @akohlmey is right that you would have issues with the validity of the `kspace` precomputations -- but you can get around those by requiring users to pre-specify a very conservative accuracy and `gewald` setting, and then crossing your fingers. You also have to make sure charges sync up correctly to accelerator packages' buffers like INTEL or KOKKOS, but that's a very distant problem.

Again, the conceptual side of things is much dicier. In your position I would be aiming to generate the electrostatic _forces_ accurately, not just charges. Unless the charges themselves were what I was interested in (and if I were, would the accuracy I gain from making them _variable_ charges outweigh the accuracy lost from making them _point_ charges in the first place?). You may also be interested in this recent paper: https://www.nature.com/articles/s41467-022-29243-2

**Comment by akohlmey** [comment]
@yuxzhou Please let us know if the information from the multiple messages is providing you with sufficient input to proceed or whether you have additional questions (and then please ask them). If there are no questions, please close the issue. Thanks.

**Comment by akohlmey** [comment]
Closing due to inactivity
